File: c++/Graph.cpp
#include <iostream>
#include <vector>
#include <map>

using namespace std;

bool vector_includes(vector<string> vec, string elem);
class Graph {
  public:
    vector<vector<int>> matrix;
    map<string, int> vertices;

    Graph(int size, bool isWeighted, bool isDirected);
    string addVertex(string vertex);
    void addEdge(string start, string end, int weight);
    void printMatrix();
    void dfs(string start, vector<string>* visited, vector<int>* arr, map<string, int>::iterator it);
    void bfs(string start, vector<int>* arr, map<string, int>::iterator it);
    void printTraversal(vector<int> arr, string text);

  private:
    int size;
    bool isWeighted;
    bool isDirected;
};

int main() {
  Graph* graph = new Graph(5, false, false);

  string a = graph->addVertex("A");
  string b = graph->addVertex("B");
  string c = graph->addVertex("C");
  string d = graph->addVertex("D");
  string e = graph->addVertex("E");

  graph->addEdge(a, b, 0);
  graph->addEdge(a, c, 0);
  graph->addEdge(b, c, 0);
  graph->addEdge(b, d, 0);
  graph->addEdge(c, e, 0);
  graph->addEdge(d, e, 0);

  cout << "Unweighted and Undirected Graph" << endl;
  graph->printMatrix();

  graph = new Graph(5, true, false);

  a = graph->addVertex("A");
  b = graph->addVertex("B");
  c = graph->addVertex("C");
  d = graph->addVertex("D");
  e = graph->addVertex("E");

  graph->addEdge(a, b, 10);
  graph->addEdge(a, c, 10);
  graph->addEdge(b, c, 5);
  graph->addEdge(b, d, 15);
  graph->addEdge(c, e, 15);
  graph->addEdge(d, e, 5);

  cout << "\nWeighted and Undirected Graph" << endl;
  graph->printMatrix();

  graph = new Graph(5, true, true);

  a = graph->addVertex("A");
  b = graph->addVertex("B");
  c = graph->addVertex("C");
  d = graph->addVertex("D");
  e = graph->addVertex("E");

  graph->addEdge(a, b, 10);
  graph->addEdge(a, c, 10);
  graph->addEdge(b, c, 5);
  graph->addEdge(b, d, 15);
  graph->addEdge(c, e, 15);
  graph->addEdge(d, e, 5);

  cout << "\nWeighted and Directed Graph" << endl;
  graph->printMatrix();

  vector<string> visited;
  vector<int> arr;
  map<string, int>::iterator it = graph->vertices.begin();

  graph->dfs(a, &visited, &arr, it);

  graph->printTraversal(arr, "\ndfs traversal");

  arr.clear();

  graph->bfs(a, &arr, it);

  graph->printTraversal(arr, "bfs traversal");

  return 0;
}

Graph::Graph(int size, bool isWeighted, bool isDirected) {
  this->size = size;
  this->isWeighted = isWeighted;
  this->isDirected = isDirected;

  for (int i = 0; i < this->size; i++) {
    vector<int> col;

    for (int j = 0; j < this->size; j++) {
      col.push_back(0);
    }

    matrix.push_back(col);
  }
}

string Graph::addVertex(string vertex) {
  if (this->vertices.size() < this->size) {
    this->vertices[vertex] = this->vertices.size();
  }

  return vertex;
}

void Graph::addEdge(string start, string end, int weight) {
  int s = this->vertices[start];
  int e = this->vertices[end];

  if (this->isWeighted) {
    this->matrix[s][e] = weight;

    if (!this->isDirected) {
      this->matrix[e][s] = weight;
    }
  } else {
    this->matrix[s][e] = 1;

    if (!this->isDirected) {
      this->matrix[e][s] = 1;
    }
  }
}

void Graph::printMatrix() {
  cout << "  ";

  for (map<string, int>::iterator it = this->vertices.begin(); it != this->vertices.end(); it++) {
    cout << " " << it->first << " ";
  }

  cout << endl;

  map<string, int>::iterator it = this->vertices.begin();

  for (int i = 0; i < 5; i++) {
    for (int j = 0; j < 5; j++) {      
      if (j == 0) {
        cout << (it++)->first << " ";
      }

      if (this->matrix[i][j] < 10) {
        cout << " " << this->matrix[i][j] << " ";
      } else {
        cout << this->matrix[i][j] << " ";
      }
    }

    cout << endl;
  }
}

void Graph::dfs(string start, vector<string>* visited, vector<int>* arr, map<string, int>::iterator it) {
  int vertex = this->vertices[start]; 
  visited->push_back(start);
  arr->push_back(vertex);

  for (int i = 0; i < this->size; i++) {
    int n = this->matrix[vertex][i];

    if (n > 0) {
      string neighbor = next(it, i)->first;

      if (!vector_includes(*visited, neighbor)) {
        this->dfs(neighbor, visited, arr, it);
      }
    }
  }
}

void Graph::bfs(string start, vector<int>* arr, map<string, int>::iterator it) {
  vector<string> visited;
  vector<string> queue;
  visited.push_back(start);
  queue.push_back(start);

  while (queue.size() > 0) {
    int vertex = this->vertices[queue[0]];

    queue.erase(queue.begin());

    arr->push_back(vertex);

    for (int i = 0; i < this->size; i++) {
      int n = this->matrix[vertex][i];

      if (n > 0) {
        string neighbor = next(it, i)->first;

        if (!vector_includes(visited, neighbor)) {
          visited.push_back(neighbor);
          queue.push_back(neighbor);
        }
      }
    }
  }
}

void Graph::printTraversal(vector<int> arr, string text) {
  cout << text << ": [";

  for (int i = 0; i < arr.size(); i++) {
    if (i == arr.size() - 1) {
      cout << arr[i] << "]\n";
    } else {
      cout << arr[i] << ", ";
    }
  }
}

bool vector_includes(vector<string> vec, string elem) {
  for (string v: vec) {
    if (v == elem) return true;
  }

  return false;
}

File: c#/Graph.cs
class Edge
{
  public Vertex start;
  public Vertex end;
  public int? weight;

  public Edge(Vertex start, Vertex end, int? weight)
  {
    this.start = start;
    this.end = end;
    this.weight = weight;
  }
}

class Vertex
{
  public string data;
  public List<Edge> edges = new List<Edge>();

  public Vertex(string data)
  {
    this.data = data;
  }

  public void addEdge(Vertex end, int? weight)
  {
    var edge = new Edge(this, end, weight);

    this.edges.Add(edge);
  }

  public void removeEdge(Vertex end)
  {
    this.edges.RemoveAll((edge) => edge.end == end);
  }
}

class Graph
{
  public List<Vertex> vertices = new List<Vertex>();
  public bool isWeighted;
  public bool isDirected;

  public Graph(bool isWeighted, bool isDirected)
  {
    this.isWeighted = isWeighted;
    this.isDirected = isDirected;
  }

  public void print()
  {
    foreach (var vertex in vertices) {
      Console.WriteLine(vertex.data);

      foreach (var edge in vertex.edges) {
        if (this.isWeighted) {
          Console.WriteLine($"{vertex.data} -> {edge.end.data}: {edge.weight}");
        } else {
          Console.WriteLine($"{vertex.data} -> {edge.end.data}");
        }
      }
    }
  }

  public void PrintTraversal(List<string> arr, string text)
  {
    Console.Write($"{text}: [");

    for (int i = 0; i < arr.Count; i++) {
      if (i == arr.Count - 1) {
        Console.Write($"{arr[i]}]\n");
      } else {
        Console.Write($"{arr[i]}, ");
      }
    }
  }

  public Vertex addVertex(string data)
  {
    var vertex = new Vertex(data);

    this.vertices.Add(vertex);

    return vertex;
  }

  public void removeVertex(Vertex vertex)
  {
    this.vertices.Remove(vertex);
  }

  public void addEdge(Vertex start, Vertex end, int? weight)
  {
    if (!this.isWeighted) {
      weight = null;
    }

    start.addEdge(end, weight);

    if (!this.isDirected) {
      end.addEdge(start, weight);
    }
  }

  public void removeEdge(Vertex start, Vertex end)
  {
    start.removeEdge(end);

    if (!this.isDirected) {
      end.removeEdge(start);
    }
  }

  public void dfs(Vertex start, List<Vertex> visited, List<string> arr)
  {
    visited.Add(start);
    arr.Add(start.data);

    foreach (Edge edge in start.edges) {
      Vertex neighbor = edge.end;

      if (!visited.Contains(neighbor)) {
        this.dfs(neighbor, visited, arr);
      }
    }
  }

  public void Bfs(Vertex start, List<string> arr)
  {
    var visited = new List<Vertex>();
    var queue = new List<Vertex>();

    visited.Add(start);
    queue.Add(start);

    while (queue.Count > 0) {
      Vertex vertex = queue[0];

      queue.RemoveAt(0);

      arr.Add(vertex.data);

      foreach (Edge edge in vertex.edges) {
        Vertex neighbor = edge.end;

        if (!visited.Contains(neighbor)) {
          visited.Add(neighbor);
          queue.Add(neighbor);
        }
      }
    }
  }
}

var graph = new Graph(false, false);

var v0 = graph.addVertex("0");
var v1 = graph.addVertex("1");
var v2 = graph.addVertex("2");
var v3 = graph.addVertex("3");
var v4 = graph.addVertex("4");

graph.addEdge(v0, v1, null);
graph.addEdge(v0, v2, null);
graph.addEdge(v1, v2, null);
graph.addEdge(v1, v3, null);
graph.addEdge(v2, v4, null);
graph.addEdge(v3, v4, null);

Console.WriteLine("Unweighted and Undirected Graph");
graph.print();

graph = new Graph(true, false);

v0 = graph.addVertex("0");
v1 = graph.addVertex("1");
v2 = graph.addVertex("2");
v3 = graph.addVertex("3");
v4 = graph.addVertex("4");

int w01 = 10;
int w02 = 10;
int w12 = 5;
int w13 = 15;
int w24 = 15;
int w34 = 5;

graph.addEdge(v0, v1, w01);
graph.addEdge(v0, v2, w02);
graph.addEdge(v1, v2, w12);
graph.addEdge(v1, v3, w13);
graph.addEdge(v2, v4, w24);
graph.addEdge(v3, v4, w34);

Console.WriteLine("\nWeighted and Undirected Graph");
graph.print();

graph = new Graph(true, true);

v0 = graph.addVertex("0");
v1 = graph.addVertex("1");
v2 = graph.addVertex("2");
v3 = graph.addVertex("3");
v4 = graph.addVertex("4");

graph.addEdge(v0, v1, w01);
graph.addEdge(v0, v2, w02);
graph.addEdge(v1, v2, w12);
graph.addEdge(v1, v3, w13);
graph.addEdge(v2, v4, w24);
graph.addEdge(v3, v4, w34);

Console.WriteLine("\nWeighted and Directed Graph");
graph.print();

var visited = new List<Vertex>();
var arr = new List<string>();

graph.dfs(v0, visited, arr);

graph.PrintTraversal(arr, "\ndfs traversal");

arr = new List<string>();

graph.Bfs(v0, arr);

graph.PrintTraversal(arr, "bfs traversal");